cpp/ql/test/query-tests/Likely Bugs/RedundantNullCheckSimple/RedundantNullCheckSimple.cpp

```diff
@@ -69,3 +69,44 @@ int test_inverted_logic(int *p) {
     return 0;
   }
 }
+
+void test_indirect_local() {
+  int a = 0;
+  int *p = &a;
+  int **pp = &p;
+  int x;
+  x = **pp;
+  if (*pp == nullptr) { // BAD
+    return;
+  }
+}
+
+void test_field_local(bool boolvar) {
+  int a = 0;
+  struct {
+    int *p;
+  } s = { &a };
+  auto sp = &s;
+
+  if (boolvar) {
+    int x = *sp->p;
+    if (sp->p == nullptr) { // BAD
+      return;
+    }
+  } else {
+    int *x = sp->p;
+    if (sp == nullptr) { // BAD [NOT DETECTED]
```

Comment:
Is this not detected because there's a `FieldAddress` instruction (with offset 0) in between the `Load` of `sp` and the `Load` of `sp->p`?

Reply:
That sounds plausible. Offset 0 is of course special, but it makes sense to handle this for all offsets. That means we can't handle it in GVN, but it'll have to be in the query. I'm willing to assume that the program is well-behaved enough that a dereference at any offset (including `p[i]`) implies that `p` isn't null.

We could put a query-specific layer on top of GVN that equates all pointers derived by pointer arithmetic from a common `ValueNumber`. That means we'd get an alert if `p + j` is checked for nullness after `p + k` is dereferenced. Such code is probably worthy of an alert.

Reply:
That's going to be quadratic, won't it? If we're careful, it's only quadratic in the number of `ValueNumber`s, but I'm nervous anyway.

Reply:
I'm proposing it as a spec, not as an implementation.